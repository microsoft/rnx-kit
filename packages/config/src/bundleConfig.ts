import type { AllPlatforms } from "@rnx-kit/tools-react-native/platform";
import type { PluginOptions as CyclicDetectorOptions } from "@rnx-kit/metro-plugin-cyclic-dependencies-detector";
import type { Options as DuplicateDetectorOptions } from "@rnx-kit/metro-plugin-duplicates-checker";
import type { OutputOptions } from "metro";

export type TypeScriptValidationOptions = {
  /**
   * Controls whether an error is thrown when type-validation fails.
   */
  throwOnError?: boolean;
};

/**
 * Parameters controlling bundler plugins.
 */
export type BundlerPlugins = {
  /**
   * Choose whether to detect cycles in the dependency graph. `true` uses defaults,
   * while `CyclicDetectorOptions` lets you control the detection process.
   */
  detectCyclicDependencies?: boolean | CyclicDetectorOptions;

  /**
   * Choose whether to detect duplicate packages in the dependency graph. `true` uses defaults,
   * while `DuplicateDetectorOptions` lets you control the detection process.
   */
  detectDuplicateDependencies?: boolean | DuplicateDetectorOptions;

  /**
   * Choose whether to type-check source files using TypeScript. `true` uses defaults,
   * while `TypeScriptValidationOptions` lets you control the validation process.
   */
  typescriptValidation?: boolean | TypeScriptValidationOptions;

  /**
   * Choose whether to enable tree shaking.
   *
   * Note that Metro ignores custom serializers (which this feature depends on)
   * when outputting RAM bundle format.
   *
   * Only applies to `rnx-bundle` command.
   */
  treeShake?: boolean;
};

/**
 * Parameters controlling how a bundle is constructed.
 */
export type BundleParameters = BundlerPlugins & {
  /**
   * Path to the .js file which is the entry-point for building the bundle.
   * Either absolute, or relative to the package.
   */
  entryFile?: string;

  /**
   * Path to the output bundle file. Either absolute or relative to the package.
   */
  bundleOutput?: string;

  /**
   * Encoding scheme to use when writing the bundle file. Currently limited
   * to UTF-8, UTF-16 (little endian), and 7-bit ASCII.
   *
   * @see https://nodejs.org/api/buffer.html#buffer_buffers_and_character_encodings
   */
  bundleEncoding?: OutputOptions["bundleEncoding"];

  /**
   * Path to use when creating the bundle source map file.
   * Either absolute, or relative to the package.
   */
  sourcemapOutput?: string;

  /**
   * Path to the package's source files. Used to make source-map paths relative and therefore portable.
   */
  sourcemapSourcesRoot?: string;

  /**
   * Controls whether or not SourceMapURL is reported as a full path or just a file name.
   */
  sourcemapUseAbsolutePath?: boolean;

  /**
   * Path where all bundle assets (strings, images, fonts, sounds, ...) are written.
   * Either absolute, or relative to the package. If not given, assets are ignored.
   */
  assetsDest?: string;

  /**
   * Force the "Indexed RAM" bundle file format, even when targeting Android.
   * For more details, see https://facebook.github.io/metro/docs/bundling.
   *
   * Only applies to the `rnx-ram-bundle` command.
   */
  indexedRamBundle?: boolean;
};

/**
 * Defines how a package is bundled. Includes shared bundling parameters with platform-specific overrides.
 */
export type BundleConfig = BundleParameters & {
  /**
   * Unique identifier for this bundle configuration. Only needed when specifying multiple bundle configurations.
   */
  id?: string;

  /**
   * The platform(s) for which this package may be bundled.
   */
  targets?: AllPlatforms[];

  /**
   * Platform-specific overrides for bundling. Any parameter not listed in an override gets
   * its value from the shared bundle configuration, or falls back to defaults.
   */
  platforms?: { [K in AllPlatforms]?: BundleParameters };
};
