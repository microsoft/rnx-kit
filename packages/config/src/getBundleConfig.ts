import type { AllPlatforms } from "@rnx-kit/tools-react-native/platform";
import castArray from "lodash/castArray";
import type { KitConfig } from "./kitConfig";
import type { BundleParameters, BundleConfig } from "./bundleConfig";

function failOnUnsupportedProp(
  parameters: BundleParameters,
  name: string,
  message: string
): void {
  if (Object.prototype.hasOwnProperty.call(parameters, name)) {
    throw new Error(
      `The bundle configuration property '${name}' is no longer supported. ${message}`
    );
  }
}

function failOnRenamedProp(
  parameters: BundleParameters,
  oldName: string,
  newName: keyof BundleParameters
): void {
  failOnUnsupportedProp(parameters, oldName, `Use '${newName}' instead.`);
}

function failOnUnsupportedProps(parameters: BundleParameters): void {
  failOnRenamedProp(parameters, "experimental_treeShake", "treeShake");

  failOnRenamedProp(parameters, "entryPath", "entryFile");
  failOnRenamedProp(parameters, "sourceMapPath", "sourcemapOutput");
  failOnRenamedProp(
    parameters,
    "sourceMapSourceRootPath",
    "sourcemapSourcesRoot"
  );
  failOnRenamedProp(parameters, "assetsPath", "assetsDest");

  failOnUnsupportedProp(
    parameters,
    "distPath",
    "You can control the bundle path and source-map path using 'bundleOutput' and 'sourcemapOutput', respectively."
  );
  failOnUnsupportedProp(
    parameters,
    "bundlePrefix",
    "You can control the bundle file name using 'bundleOutput'."
  );
}

/**
 * Get a bundle configuration from the rnx-kit configuration.
 *
 * If an id is given, search for the matching bundle definition. Otherwise, use the first bundle definition.
 *
 * @param config The package's rnx-kit configuration
 * @param id Optional identity of the target bundle configuration
 * @returns Bundle configuration, or `undefined` if nothing was found.
 */
export function getBundleConfig(
  config: KitConfig,
  id?: string
): BundleConfig | undefined {
  if (!config.bundle) {
    return undefined;
  }

  // 5/20/2022: fail when `bundle` is set to true, which is from the old config format
  //            remove on the next 0.x bump or when going to 1.0, whichever comes first
  // @ts-expect-error --^
  if (config.bundle === true) {
    throw new Error(
      `The rnx-kit configuration property 'bundle' no longer supports boolean values. Bundling is always enabled with sensible defaults. You should remove the 'bundle' property to make use of the defaults, or specify the bundle configuration as an object.`
    );
  }

  const bundles = castArray(config.bundle);

  let bundle: BundleConfig | undefined;
  if (id) {
    bundle = bundles.find((b) => b.id === id);
  } else {
    bundle = bundles.length > 0 ? bundles[0] : undefined;
  }
  if (!bundle) {
    return undefined;
  }

  // 5/20/2022: fail when there are signs of the old config format
  //            remove on the next 0.x bump or when going to 1.0, whichever comes first
  failOnUnsupportedProps(bundle);

  return bundle;
}

/**
 * Resolves the platform selector for a bundle configuration
 *
 * @param bundle Bundle config to resolve (includes the optional platform selectors)
 * @param platform Target platform
 * @returns Bundle config containing platform-specific overrides
 */
export function getPlatformBundleConfig(
  bundle: BundleConfig,
  platform: AllPlatforms
): BundleConfig {
  const platformValues = bundle.platforms
    ? bundle.platforms[platform]
    : undefined;
  return platformValues ? { ...bundle, ...platformValues } : bundle;
}
