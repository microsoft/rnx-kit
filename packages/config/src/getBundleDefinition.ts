import type { AllPlatforms } from "@rnx-kit/tools-react-native/platform";
import castArray from "lodash/castArray";
import type { KitConfig } from "./kitConfig";
import type { BundleDefinition } from "./bundleConfig";

function failOnUnsupportedDefinitionProp(
  o: unknown,
  name: string,
  message: string
): void {
  if (Object.prototype.hasOwnProperty.call(o, name)) {
    throw new Error(
      `The '${name}' configuration property is no longer supported. ${message}`
    );
  }
}

function failOnRenamedDefinitionProp(
  o: unknown,
  oldName: string,
  newName: string
): void {
  failOnUnsupportedDefinitionProp(o, oldName, `Use '${newName}' instead.`);
}

function failOnUnsupportedDefinitionProps(bundle: BundleDefinition): void {
  failOnRenamedDefinitionProp(bundle, "experimental_treeShake", "treeShake");

  failOnRenamedDefinitionProp(bundle, "entryPath", "entryFile");
  failOnRenamedDefinitionProp(bundle, "sourceMapPath", "sourcemapOutput");
  failOnRenamedDefinitionProp(
    bundle,
    "sourceMapSourceRootPath",
    "sourcemapSourcesRoot"
  );
  failOnRenamedDefinitionProp(bundle, "assetsPath", "assetsDest");

  failOnUnsupportedDefinitionProp(
    bundle,
    "distPath",
    "You can control the bundle path and source-map path using 'bundleOutput' and 'sourcemapOutput', respectively."
  );
  failOnUnsupportedDefinitionProp(
    bundle,
    "bundlePrefix",
    "You can control the bundle file name using 'bundleOutput'."
  );
}

/**
 * Get a bundle definition from the rnx-kit configuration.
 *
 * If an id is given, search for the matching bundle definition. Otherwise, use the first bundle definition.
 *
 * @param config rnx-kit configuration
 * @param id Optional identity of the target bundle definition to return
 * @returns Bundle definition
 */
export function getBundleDefinition(
  config: KitConfig,
  id?: string
): BundleDefinition {
  // 'bundle' property not set?
  if (!Object.prototype.hasOwnProperty.call(config, "bundle")) {
    throw new Error(
      "Bundling is not enabled in the rnx-kit configuration for this package."
    );
  }

  // 'bundle' property explicitly set to null/undefined/false?
  if (
    config.bundle === undefined ||
    config.bundle === null ||
    config.bundle === false
  ) {
    throw new Error(
      "Bundling is explicitly disabled in the rnx-kit configuration for this package."
    );
  }

  // default bundle config?
  if (config.bundle === true) {
    return {}; // empty -> default config
  }

  const bundles = castArray(config.bundle);

  let bundle: BundleDefinition | undefined;
  if (id) {
    bundle = bundles.find((b) => b.id === id);
    if (!bundle) {
      throw new Error(
        `Bundle definition with id '${id}' was not found in the rnx-kit configuration for this package.`
      );
    }
  } else {
    bundle = bundles.length > 0 ? bundles[0] : undefined;
    if (!bundle) {
      throw new Error(
        "No bundle definitions were found in the rnx-kit configuration for this package."
      );
    }
  }

  // 5/20/2022: fail when there are signs of the old config format
  //            remove on the next 0.x bump or when going to 1.0, whichever comes first
  failOnUnsupportedDefinitionProps(bundle);

  return bundle;
}

/**
 * Resolves the platform selector for a bundle definition
 *
 * @param bundle Bundle definition to resolve (includes the optional platform selectors)
 * @param platform Target platform
 * @returns Bundle definition containing platform-specific overrides
 */
export function getBundlePlatformDefinition(
  bundle: BundleDefinition,
  platform: AllPlatforms
): BundleDefinition {
  const platformValues = bundle.platforms
    ? bundle.platforms[platform]
    : undefined;
  return platformValues ? { ...bundle, ...platformValues } : bundle;
}
