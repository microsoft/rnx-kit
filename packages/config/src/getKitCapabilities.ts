import { warn } from "@rnx-kit/console";
import semver from "semver";
import type { KitConfig } from "./kitConfig";

type RequiredConfig = Required<
  Pick<
    KitConfig,
    "capabilities" | "kitType" | "reactNativeVersion" | "reactNativeDevVersion"
  >
>;

type OptionalConfig = Pick<KitConfig, "customProfiles">;

export type KitCapabilities = RequiredConfig & OptionalConfig;

export function getKitCapabilities({
  capabilities = [],
  kitType = "library",
  reactNativeVersion,
  reactNativeDevVersion: rawDevVersion,
  customProfiles,
}: KitConfig): KitCapabilities {
  if (
    !reactNativeVersion ||
    (!semver.valid(reactNativeVersion) &&
      !semver.validRange(reactNativeVersion))
  ) {
    throw new Error(`'${reactNativeVersion}' is not a valid version range`);
  }

  if (kitType === "app" && rawDevVersion) {
    warn("'reactNativeDevVersion' is not used when 'kitType' is 'app'");
  }

  const reactNativeDevVersion =
    rawDevVersion || semver.minVersion(reactNativeVersion)?.version;

  if (
    !reactNativeDevVersion ||
    !semver.subset(reactNativeDevVersion, reactNativeVersion)
  ) {
    throw new Error(
      `'${reactNativeDevVersion}' is not a valid dev version because it does not satisfy supported version range '${reactNativeVersion}'`
    );
  }

  return {
    capabilities,
    kitType,
    reactNativeVersion,
    reactNativeDevVersion,
    customProfiles,
  };
}
