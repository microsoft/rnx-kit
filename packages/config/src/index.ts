export type {
  BundleConfig,
  BundleDefinition,
  BundleParameters,
  BundlerRuntimeParameters,
  TypeScriptValidationOptions,
} from "./bundleConfig";

export {
  getBundleDefinition,
  getBundlePlatformDefinition,
} from "./getBundleDefinition";

export { getKitCapabilities } from "./getKitCapabilities";
export type { KitCapabilities } from "./getKitCapabilities";

export { getKitConfig } from "./getKitConfig";
export type { GetKitConfigOptions } from "./getKitConfig";

export { getServerConfig } from "./getServerConfig";

export type {
  Capability,
  DependencyVersions,
  GetDependencyVersions,
  KitConfig,
  KitType,
  MetaCapability,
} from "./kitConfig";

export type { ServerConfig } from "./serverConfig";
