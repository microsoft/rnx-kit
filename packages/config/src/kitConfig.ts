import type { BundleConfig } from "./bundleConfig";
import type { ServerConfig } from "./serverConfig";

export type MetaCapability = "core/testing";

export type Capability =
  | MetaCapability
  | "core"
  | "core-android"
  | "core-ios"
  | "core-macos"
  | "core-windows"
  | "animation"
  | "babel-preset-react-native"
  | "base64"
  | "checkbox"
  | "clipboard"
  | "datetime-picker"
  | "filesystem"
  | "floating-action"
  | "gestures"
  | "hermes"
  | "hooks"
  | "html"
  | "jest"
  | "lazy-index"
  | "masked-view"
  | "metro"
  | "metro-config"
  | "metro-core"
  | "metro-react-native-babel-transformer"
  | "metro-resolver"
  | "metro-runtime"
  | "modal"
  | "navigation/native"
  | "navigation/stack"
  | "netinfo"
  | "popover"
  | "react"
  | "react-dom"
  | "react-test-renderer"
  | "safe-area"
  | "screens"
  | "shimmer"
  | "sqlite"
  | "storage"
  | "svg"
  | "test-app"
  | "webview";

export type DependencyVersions = Record<string, string>;

export type GetDependencyVersions = () => DependencyVersions;

export type KitType = "app" | "library";

/**
 * Configuration information for an rnx-kit package. This is retrieved from 'rnx-kit' in package.json.
 */
export type KitConfig = {
  /**
   * Whether this kit is an "app" or a "library".
   * @default "library"
   */
  kitType?: KitType;

  /**
   * Supported versions of React Native. Must be parseable by
   * [node-semver](https://github.com/npm/node-semver).
   */
  reactNativeVersion?: string;

  /**
   * The version of React Native to use for development. Must be parseable by
   * [node-semver](https://github.com/npm/node-semver). If omitted, the minimum
   * supported version will be used.
   * @default minVersion(reactNativeVersion)
   */
  reactNativeDevVersion?: string;

  /**
   * Configures how `align-deps` should align dependencies for this package.
   */
  alignDeps?: {
    /**
     * Presets to use for aligning dependencies.
     * @default ["microsoft/react-native"]
     */
    presets?: string[];

    /**
     * Requirements for this package, e.g. `react-native@>=0.66`.
     */
    requirements?: string[] | { development: string[]; production: string[] };

    /**
     * Capabilities used by the kit.
     * @default []
     */
    capabilities?: Capability[];
  };

  /**
   * Specifies how the package is bundled.
   */
  bundle?: BundleConfig | BundleConfig[];

  /**
   * Specifies how the package's bundle server is configured.
   */
  server?: ServerConfig;

  /**
   * Capabilities used by the kit.
   * @default []
   */
  capabilities?: Capability[];

  /**
   * Path to custom profiles. This can be a path to a JSON file, a `.js` file,
   * or a module name. The module must default export an object similar to the
   * one below.
   *
   *     module.exports = {
   *       "0.63": {
   *         "my-capability": {
   *           "name": "my-module",
   *           "version": "1.0.0",
   *         },
   *       },
   *       "0.64": {
   *         "my-capability": {
   *           "name": "my-module",
   *           "version": "1.1.0",
   *         },
   *       },
   *     };
   *
   * For a more complete example, please take a look at the default profiles:
   * https://github.com/microsoft/rnx-kit/blob/769e9fa290929effd5111884f1637c21326b5a95/packages/dep-check/src/profiles.ts#L11
   */
  customProfiles?: string;
};
