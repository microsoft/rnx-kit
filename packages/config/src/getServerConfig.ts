import type { KitConfig } from "./kitConfig";
import type { ServerConfig } from "./serverConfig";
import { pickValues } from "@rnx-kit/tools-language";
import { getBundleDefinition } from "./getBundleDefinition";

/**
 * Get server configuration from the rnx-kit configuration.
 *
 * @param config rnx-kit configuration
 * @returns Server configuration
 */
export function getServerConfig(config: KitConfig): ServerConfig {
  // 'server' property not set?
  if (!Object.prototype.hasOwnProperty.call(config, "server")) {
    // if bundling is enabled, use the bundle definition to control the server
    try {
      const bundleDefinition = getBundleDefinition(config);
      return {
        ...pickValues(bundleDefinition, [
          "detectCyclicDependencies",
          "detectDuplicateDependencies",
          "typescriptValidation",
          "treeShake",
        ]),
      };
    } catch {
      throw new Error(
        "Bundle serving is not enabled. The rnx-kit configuration for this package has no server config, nor does it have bundling config to use as a baseline for running the bundle server."
      );
    }
  }

  // 'server' property explicitly set to undefined?
  if (config.server === undefined) {
    throw new Error(
      "Bundle serving is explicitly disabled in the rnx-kit configuration for this package."
    );
  }

  return config.server;
}
