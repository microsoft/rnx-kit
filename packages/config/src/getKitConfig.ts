import path from "path";
import { readPackage } from "@rnx-kit/tools-node";
import { KitConfig } from "./kitConfig";

/**
 * Options for retrieving a kit config. The default is equivalent to passing { cwd: process.cwd() }
 */
export type GetKitConfigOptions = {
  /**
   * Retrieve the rnx-kit configuration from a package module.
   */
  module?: string;

  /**
   * Retrive the rnx-kit configuration from a directory containing 'package.json'.
   */
  cwd?: string;
};

/**
 * Query for a package's rnx-kit configuration.
 *
 * @param options Options for retrieving the configuration.
 * @returns
 */
export function getKitConfig(
  options: GetKitConfigOptions = {}
): KitConfig | undefined {
  // find the package dir that holds the rnx-kit configuration
  const packageDir = options.module
    ? path.dirname(require.resolve(options.module + "/package.json"))
    : options.cwd || process.cwd();

  // try to read package.json
  try {
    const packageJson = readPackage(packageDir);
    return packageJson["rnx-kit"];
  } catch {
    return undefined;
  }
}
