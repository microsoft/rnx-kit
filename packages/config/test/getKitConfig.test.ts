import { getKitConfig } from "../src/getKitConfig";
import path from "path";

describe("getKitConfig()", () => {
  const currentWorkingDir = process.cwd();

  function fixturePath(): string {
    return path.join(currentWorkingDir, "test", "__fixtures__");
  }

  function packagePath(name: string): string {
    return path.join(fixturePath(), "node_modules", name);
  }

  afterEach(() => process.chdir(currentWorkingDir));

  test("returns undefined for an unconfigured package when using the current working directory", () => {
    process.chdir(packagePath("kit-test-unconfigured"));
    expect(getKitConfig()).toBeUndefined();
  });

  test("returns undefined for an unconfigured package when using an explicit working directory", () => {
    expect(
      getKitConfig({ cwd: packagePath("kit-test-unconfigured") })
    ).toBeUndefined();
  });

  test("returns undefined for an unconfigured package when using a module name", () => {
    expect(getKitConfig({ module: "kit-test-unconfigured" })).toBeUndefined();
  });

  test("returns rnx-kit configuration when using the current working directory", () => {
    process.chdir(packagePath("kit-test-configured"));
    expect(getKitConfig()).toMatchSnapshot();
  });

  test("returns rnx-kit configuration when using an explicit working directory", () => {
    expect(
      getKitConfig({ cwd: packagePath("kit-test-configured") })
    ).toMatchSnapshot();
  });

  test("returns rnx-kit configuration when using a module name", () => {
    expect(getKitConfig({ module: "kit-test-configured" })).toMatchSnapshot();
  });
});
