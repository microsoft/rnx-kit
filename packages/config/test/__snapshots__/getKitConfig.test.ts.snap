// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getKitConfig() returns rnx-kit configuration when using a module name 1`] = `
Object {
  "bundle": Object {
    "bundleOutput": "./app.bundle",
    "entryFile": "./core-entry.js",
    "id": "core",
    "platforms": Object {
      "android": Object {
        "assetsDest": "./build-out/res",
      },
    },
    "targets": Array [
      "ios",
      "android",
      "macos",
      "windows",
    ],
  },
}
`;

exports[`getKitConfig() returns rnx-kit configuration when using an explicit working directory 1`] = `
Object {
  "bundle": Object {
    "bundleOutput": "./app.bundle",
    "entryFile": "./core-entry.js",
    "id": "core",
    "platforms": Object {
      "android": Object {
        "assetsDest": "./build-out/res",
      },
    },
    "targets": Array [
      "ios",
      "android",
      "macos",
      "windows",
    ],
  },
}
`;

exports[`getKitConfig() returns rnx-kit configuration when using the current working directory 1`] = `
Object {
  "bundle": Object {
    "bundleOutput": "./app.bundle",
    "entryFile": "./core-entry.js",
    "id": "core",
    "platforms": Object {
      "android": Object {
        "assetsDest": "./build-out/res",
      },
    },
    "targets": Array [
      "ios",
      "android",
      "macos",
      "windows",
    ],
  },
}
`;
